File: src/nodejs/fastify/src/models/ComplexOrderResult.js
import { z } from 'zod';

// Mirrors UserOrderStats in contracts/grpc/benchmark.proto. Wire names are
// camelCase. See contracts/rest/canonical-payloads.md.
export const ComplexOrderResultSchema = z.object({
  userId: z.number(),
  userName: z.string(),
  totalOrders: z.number(),
  totalValue: z.number(),
  averageOrderValue: z.number()
});

export function createComplexOrderResult(row) {
  return {
    userId: parseInt(row.userId),
    userName: row.userName,
    totalOrders: parseInt(row.totalOrders),
    totalValue: parseFloat(row.totalValue),
    averageOrderValue: parseFloat(row.averageOrderValue)
  };
}

export function validateComplexOrderResult(data) {
  return ComplexOrderResultSchema.parse(data);
}

File: src/dart/vaden/lib/models/complex_order_result.dart
class ComplexQueryResult {
  final int periodDays;
  final int totalUsers;
  final List<UserStats> data;

  const ComplexQueryResult({
    required this.periodDays,
    required this.totalUsers,
    required this.data,
  });

  Map<String, dynamic> toJson() {
    return {
      'periodDays': periodDays,
      'totalUsers': totalUsers,
      'data': data.map((e) => e.toJson()).toList(),
    };
  }
}

class UserStats {
  final int userId;
  final String userName;
  final int totalOrders;
  final double totalValue;
  final double averageOrderValue;

  const UserStats({
    required this.userId,
    required this.userName,
    required this.totalOrders,
    required this.totalValue,
    required this.averageOrderValue,
  });

  Map<String, dynamic> toJson() {
    return {
      'userId': userId,
      'userName': userName,
      'totalOrders': totalOrders,
      'totalValue': totalValue,
      'averageOrderValue': averageOrderValue,
    };
  }
}
